### PR TITLE
[REEF-604] Support Hadoop security tokens in YARN runtime during job submission and container creation

### PR DESCRIPTION
**Pull request by anupam128**
This change addressed the issue by:
 * Creating SecurityTokenProvider which provides security tokens
 * Add default implementation of SecurityTokenProvider, reading security token from user credentials
 * Get tokens from SecurityTokenProvider and add the tokens ContainerLaunchContext during application and container creation.

JIRA:
  [REEF-604](https://issues.apache.org/jira/browse/REEF-604)

### CODE REVIEW

File: lang/cs/.nuget/NuGet.Config [hunk 1]
```diff
@@ -21,11 +21,7 @@ under the License.
...
     <add key="disableSourceControlIntegration" value="true" />
   </solution>
-  <packageSources>
-      <add key="NuGet official package source" value="https://nuget.org/api/v2/" />
-  </packageSources>
-  <disabledPackageSources />
   <activePackageSource>
-      <add key="All" value="(Aggregate source)"  />
+      <add key="All" value="(Aggregate source)" />
   </activePackageSource>
```
Comment (anupam128) [inline lang/cs/.nuget/NuGet.Config]: Will revert.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/SecurityTokenProvider.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Provides security token for setting up YARN container context.
+ */
+@DefaultImplementation(UserCredentialSecurityTokenProvider.class)
+public interface SecurityTokenProvider {
+
+  /**
+   * Returns a ByteBuffer containing security tokens.
+   * @return a ByteBuffer
+   */
+  ByteBuffer getTokens();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/SecurityTokenProvider.java]: Is `ByteBuffer` the right type for these? I'd have expected `String` or `byte[]`?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  private static final Logger LOG = Logger.getLogger(UserCredentialSecurityTokenProvider.class.getName());
+
+  @Inject
+  public UserCredentialSecurityTokenProvider(){}
+
+  @Override
+  public ByteBuffer getTokens() {
+    try {
+      UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: `final`

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java [hunk 4]
```diff
@@ -63,16 +71,25 @@ private static Configuration getDriverConfiguration() {
...
   }
 
+  private static void setupUserToken() throws IOException {
+    UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
+    byte[] key = Files.readAllBytes(Paths.get("D:\\anupam\\AzureSystemJmKeys"));
+    Token token = new Token(key, "none".getBytes(), KIND, KIND);
+    ugi.addToken(token);
+  }
+
   /**
```
Comment (anupam128) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java]: added for testing will revert.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import javax.inject.Inject;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Reads security token from user credentials.
+ */
+public class UserCredentialSecurityTokenProvider implements SecurityTokenProvider {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: `public final`, or, better yet, just `final`.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  private static final Logger LOG = Logger.getLogger(UserCredentialSecurityTokenProvider.class.getName());
+
+  @Inject
+  public UserCredentialSecurityTokenProvider(){}
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: If this is only called via Tang, make it `private`

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Override
+  public ByteBuffer getTokens() {
+    try {
+      UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
+      Credentials credentials = ugi.getCredentials();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: `final`

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      if (credentials.numberOfTokens() > 0) {
+        DataOutputBuffer dob = new DataOutputBuffer();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: `final`. ALso, doesn't this also need to be `.close()`ed?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        credentials.writeTokenStorageToStream(dob);
+        return ByteBuffer.wrap(dob.getData(), 0, dob.getLength());
+      }
+    } catch (IOException e) {
+      LOG.log(Level.WARNING, "Could not access tokens in user credentials.", e);
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: Is this recoverable? Shouldn't we die here?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  public ByteBuffer getTokens() {
+    try {
+      UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
+      Credentials credentials = ugi.getCredentials();
+      if (credentials.numberOfTokens() > 0) {
+        DataOutputBuffer dob = new DataOutputBuffer();
+        credentials.writeTokenStorageToStream(dob);
+        return ByteBuffer.wrap(dob.getData(), 0, dob.getLength());
+      }
+    } catch (IOException e) {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: `final`

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 11]
```diff
@@ -194,8 +197,9 @@ public void submit() throws IOException, YarnException {
           " since the max application submissions is 1. Proceeding to submit application...");
     }
 
-    this.applicationSubmissionContext.setAMContainerSpec(YarnTypes.getContainerLaunchContext(launchCommand,
-        this.resources));
+    ContainerLaunchContext containerLaunchContext = YarnTypes.getContainerLaunchContext(
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: `final`

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YARNResourceLaunchHandler.java [hunk 12]
```diff
@@ -84,7 +89,9 @@ public void onNext(final ResourceLaunchEvent resourceLaunchEvent) {
             new Object[]{containerId, StringUtils.join(command, ' '), localResources});
       }
 
-      final ContainerLaunchContext ctx = YarnTypes.getContainerLaunchContext(command, localResources);
+      ByteBuffer securityTokensBuffer = this.tokenProvider.getTokens();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YARNResourceLaunchHandler.java]: `final`

File: lang/cs/.nuget/NuGet.Config [hunk 13]
```diff
@@ -21,11 +21,7 @@ under the License.
   <solution>
     <add key="disableSourceControlIntegration" value="true" />
   </solution>
-  <packageSources>
```
Comment (markusweimer) [inline lang/cs/.nuget/NuGet.Config]: Why are there changes in this file at all? It seems orthogonal to the PR

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java [hunk 14]
```diff
@@ -39,6 +46,7 @@
  */
 public final class HelloReefYarnTcp {
 
+  private static final Text KIND = new Text("CaboSystemSecurityKeyToken");
```
Comment (markusweimer) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java]: This leaks information :-)

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java [hunk 15]
```diff
@@ -63,16 +71,25 @@ private static Configuration getDriverConfiguration() {
   private static Configuration getRuntimeConfiguration(
       final int tcpBeginPort,
       final int tcpRangeCount,
-      final int tcpTryCount) {
+      final int tcpTryCount,
+      final String jobSubmissionDirectoryPrefix) {
```
Comment (markusweimer) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java]: This is a fix for another issue, right?

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java [hunk 16]
```diff
@@ -82,11 +99,16 @@ private static Configuration getRuntimeConfiguration(
...
+  public static final String DEFAULT_JOB_SUBMISSION_DIR_PREFIX = "/vol1/tmp";
+  public static void main(final String[] args) throws InjectionException, IOException {
     final int tcpBeginPort = args.length > 0 ? Integer.valueOf(args[0]) : DEFAULT_TCP_BEGIN_PORT;
     final int tcpRangeCount = args.length > 1 ? Integer.valueOf(args[1]) : DEFAULT_TCP_RANGE_COUNT;
     final int tcpTryCount = args.length > 2 ? Integer.valueOf(args[2]) : DEFAULT_TCP_RANGE_TRY_COUNT;
-    Configuration runtimeConfiguration = getRuntimeConfiguration(tcpBeginPort, tcpRangeCount, tcpTryCount);
+    final String jobSubmissionDirectoryPrefix = args.length > 3 ? args[3] : DEFAULT_JOB_SUBMISSION_DIR_PREFIX;
+
+    setupUserToken();
+    Configuration runtimeConfiguration = getRuntimeConfiguration(
```
Comment (markusweimer) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java]: `final`

**Comment by markusweimer** [comment]
@anupam128 I've done a first pass

File: lang/cs/.nuget/NuGet.Config (hunk 13 above)
Comment (anupam128) [inline lang/cs/.nuget/NuGet.Config]: Yes. Was auto added by VS2015 when I was trying to run tests. I am reverting this change.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java (hunk 9 above)
Comment (anupam128) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/UserCredentialSecurityTokenProvider.java]: This is recoverable for clusters where authentication is not required.

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java (hunk 15 above)
Comment (anupam128) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloReefYarnTcp.java]: No. But I will revert this for now.

**Comment by anupam128** [comment]
Addressed the comments

**Comment by markusweimer** [comment]
@anupam128 Thanks! I will do another pass.

**Comment by markusweimer** [comment]
Looks good, will merge.
